graveler/graveler_test.go

```diff
@@ -432,18 +432,225 @@ func TestGraveler_CreateBranch(t *testing.T) {
 func TestGraveler_UpdateBranch(t *testing.T) {
 	gravel := graveler.NewGraveler(nil,
 		&stagingMock{ValueIterator: newMockValueIterator([]graveler.ValueRecord{{Key: graveler.Key("foo/one"), Value: &graveler.Value{}}})},
-		&mockRefs{branch: &graveler.Branch{}},
+		&refsMock{branch: &graveler.Branch{}},
 	)
 	_, err := gravel.UpdateBranch(context.Background(), "", "", "")
 	if !errors.Is(err, graveler.ErrConflictFound) {
 		t.Fatal("expected update to fail on conflict")
 	}
 	gravel = graveler.NewGraveler(nil,
 		&stagingMock{ValueIterator: newMockValueIterator([]graveler.ValueRecord{})},
-		&mockRefs{branch: &graveler.Branch{}},
+		&refsMock{branch: &graveler.Branch{}},
 	)
 	_, err = gravel.UpdateBranch(context.Background(), "", "", "")
 	if err != nil {
 		t.Fatal("did not expect to get error")
 	}
 }
+
+func TestGraveler_branchLock(t *testing.T) {
+	bl := graveler.NewBranchLock()
+	closeWrite, err := bl.RequestWrite("a", defaultBranchID)
+	if err != nil {
+		t.Fatal("expected can write")
+	}
+
+	closeWrite()
+	closeWrite, err = bl.RequestWrite("a", defaultBranchID)
+	if err != nil {
+		t.Fatal("expected can write")
+	}
+	writeWhileCommitEnded := make(chan interface{})
+	commitCalled := make(chan interface{})
+	commitStarted := make(chan interface{})
+	commitEnded := make(chan interface{})
+
+	go func() {
+		close(commitCalled)
+		closeCommit, err := bl.RequestMetadataUpdate("a", defaultBranchID)
+		close(commitStarted)
+		if err != nil {
+			t.Error(err)
+		}
+		<-writeWhileCommitEnded
+		closeCommit()
+		close(commitEnded)
+	}()
+	<-commitCalled
+	// try to commit while commit is pending
+	_, err = bl.RequestWrite("a", defaultBranchID)
```

Comment:
you can't assume that you got the goroutine into the point passed metadata update